src/root.zig

```diff
@@ -300,7 +335,7 @@ pub fn AlgorithmType(
         ) void {
             var prev: ElType = 0;
             for (1.., haystack) |i, h| {
-                self.role_bonus[i] = funcTable.bonus(ctx, scores, prev, h);
+                self.role_bonus[i] = funcTable.bonusFunc(&self.impl, scores, prev, h, self.allocator);
```

Comment:
There's no need to pass an allocator here, since the `ctx` owns the Algorithm, so can access the allocator in the function table using `self.alg.allocator`.

Reply:
This applies to a number of places throughout the changes.

Reply:
(ah these must be WIP changes, as there is no `self.impl` anymore either)